### PR TITLE
[1.4] Fixed and reworked ModGore

### PR DESCRIPTION
**Pull request by Mirsario**
This PR does the following:
- Fixes ModGore, as it just hasn't been working for some reason I don't even know.
- Splits ModGore into GoreLoader, improve some code quality in general.
- Turns ModGore into a ModTexturedType. Reduced the amount of hardcode in autoloading.
- Changes how Texture-only mod gores load. Now, for every such gore, a dummy ModGore is registered with ModName/TextureName FullName. This means that gores can now be gotten using Mod/ModContent.Find/TryFind, which do not use texture paths.
- Mod.AddGore is replaced with GoreLoader.AddGoreFromTexture, which does what you would think, i.e. what I described above.
- Removes by-texture-path GetGoreSlot methods, for less hardcode and way less confusion. Content devs seem to support the switch from texture paths to ModName/ContentName.
- Renamed Gore.modGore to Gore.ModGore. We did plenty of naming convention fixes like this, should just do it for everything on the main 1.4 branch.
- Splits tML's added fields of Gore.cs into Gore.TML.cs.
- Fixes tML's added fields not being reset in Gore.NewGore.
- Temporarily added ModContent.GoreType<T>(). I would like to replace all such methods with a generalized one.

### CODE REVIEW

File: patches/tModLoader/Terraria/ModLoader/GoreLoader.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		}
+
+		//in Terraria.Main.DrawGore and DrawGoreBehind replace type checks with this
+		internal static bool DrawBackGore(Gore gore) {
+			if (gore.ModGore != null) {
+				return gore.ModGore.DrawBehind(gore);
+			}
+
+			return (((gore.type >= 706 && gore.type <= 717) || gore.type == 943 || gore.type == 1147 || (gore.type >= 1160 && gore.type <= 1162)) && (gore.frame < 7 || gore.frame > 9));
+		}
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/GoreLoader.cs:116]: Bad patch I didn't make, will need to fix it later.

File: patches/tModLoader/Terraria/ModLoader/GoreLoader.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+			}
+
+			foreach (var pair in gores) {
+				TextureAssets.Gore[pair.Key] = ModContent.GetTexture(pair.Value.Texture);
+			}
+		}
+
+		internal static void Unload() {
+			gores.Clear();
+			gores.Clear();
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/GoreLoader.cs:77]: hmm

File: patches/tModLoader/Terraria/ModLoader/GoreLoader.cs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+			GoreCount = GoreID.Count;
+		}
+
+		//in Terraria.Gore add modGore property (internal set)
+		//in Terraria.Gore.NewGore after resetting properties call ModGore.SetupGore(Main.gore[num]);
+		internal static void SetupGore(Gore gore) {
+			gore.ResetNewFields();
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/GoreLoader.cs:84]: I'd prefer this be called from `Gore` and be private.

File: patches/tModLoader/Terraria/ModLoader/GoreLoader.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+			if (gores.ContainsKey(gore.type)) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/GoreLoader.cs:86]: maybe this whole method should just return `ModGore` and use a `TryGetValue`? Then in `Gore` itself you can call `gore.ModGore?.OnSpawn(this);`

File: patches/tModLoader/Terraria/ModLoader/ModGore.cs [hunk 5]
```diff
@@ -1,117 +1,39 @@
...
-		}
+		public int Type { get; internal set; }
 
-		internal static void SetupUpdateType(Gore gore) {
-			if (gore.modGore != null && gore.modGore.updateType > 0) {
-				gore.realType = gore.type;
-				gore.type = gore.modGore.updateType;
-			}
-		}
+		public override string Name => nameOverride ?? base.Name;
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/ModGore.cs:18]: can we do a `ModGoreSimple` subclass instead?

**Review by Chicken-Bones — COMMENTED** [review]
fantastic

File: patches/tModLoader/Terraria/ModLoader/ModGore.cs (hunk 5 above)
Comment (Mirsario) [reply patches/tModLoader/Terraria/ModLoader/ModGore.cs:18]: GoreLoader.AddGoreFromTexture<T> is currently usable with custom ModGores too.

Comment (Chicken-Bones) [reply patches/tModLoader/Terraria/ModLoader/ModGore.cs:18]: good point

**Pull request merged**